### PR TITLE
Message Bus - Basic Framework

### PR DESCRIPTION
**Pull request by selvakumaar5496**

### CODE REVIEW

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/message_bus/error.py [hunk 1]
```diff
@@ -15,120 +15,24 @@
...
+    """ Generic Exception with error code and output """
 
-class MessagebusError(BaseError):
-    """
-    Parent class for the Messagebus error classes
-    """
+    def __init__(self, rc, message, *args):
+        self._rc = rc
+        self._desc = message % (args)
+        print(self._desc)
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/error.py:25]: Please remove this print.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            if name != 'MessageBroker' and name.endswith("Broker"):
+                if broker_type == cls.name:
+                    message_broker = cls(broker_conf)
+                    MessageBrokerFactory._brokers[broker_type] = message_broker
+                    return message_broker
+
+        raise MessageBusError(errno.EINVAL, "Invalid broker type %s.", \
+            broker_type)
+
+class MessageBroker:git
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker.py:45]: Why ":git" ?

File: py-utils/src/utils/message_bus/message_broker.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                    return message_broker
+
+        raise MessageBusError(errno.EINVAL, "Invalid broker type %s.", \
+            broker_type)
+
+class MessageBroker:git 
+    """ A common interface of Message Brokers"""
+
+    def __init__(self, broker_conf):
+        # TODO: Handle ports
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker.py:49]: Remove TODO

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        kafka_conf = {}
+        kafka_conf['bootstrap.servers'] = self._servers
+        kafka_conf['client.id'] = client_conf['client_id']
+
+        if client_type == 'producer':
+            producer = Producer(**kafka_conf)
+            self._clients[client_type][client_conf['client_id']] = producer
+
+        else:
+            for entry in ['offset', 'consumer_group', 'message_type', 'auto_ack']:
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:66]: add group.id to the list for validation.

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+            kafka_conf['enable.auto.commit'] = client_conf['auto_ack']
+            kafka_conf['auto.offset.reset'] = client_conf['offset']
+            kafka_conf['group.id'] = client_conf['consumer_group']
+
+            consumer = Consumer(**kafka_conf)
+            consumer.subscribe(client_conf['message_type'])
+            self._clients[client_type][client_conf['client_id']] = consumer
+
+    def send(self, message_type: str, method: str, messages: list, client_id: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:79]: Call it producer_id and as first parameter
```
    def send(self, producer_id: str, message_type: str, method: str, messages: list):
```

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        if producer is None:
+            raise MessageBusError(errno.EINVAL, "init_client is not called")
+
+        for message in messages:
+            producer.produce(message_type, bytes(message, 'utf-8'))
+            if method == 'sync':
+                producer.flush()
+        producer.flush()
+
+    def receive(self, client_id, timeout=0.5) -> list:
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:91]:
```
    def receive(self, consumer_id: str, timeout=0.5) -> list:
```

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            sys.stderr.write('%% Aborted by user\n')
+
+    def ack(self):
+        """ To manually commit offset """
+        consumer = self._clients['consumer']
+        if consumer is None:
+            raise MessageBusError(errno.EINVAL, "init_client is not called")
+        consumer.commit()
+        consumer.close()
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:119]: Remove this blank line.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/error.py (hunk 1 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/error.py:25]: happened while rebasing

File: py-utils/src/utils/message_bus/message_broker.py (hunk 2 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker.py:45]: happened while rebasing

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 7 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:119]: cleared

File: py-utils/src/utils/message_bus/message_broker.py (hunk 3 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker.py:49]: Removed

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 4 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:66]: validated

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 5 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:79]: added as first parameter

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 6 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:91]: renamaed to consumer_id

File: py-utils/src/utils/message_bus/message_broker.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        raise MessageBusError(errno.EINVAL, "Invalid broker type %s.", \
+            broker_type)
+
+
+class MessageBroker:
+    """ A common interface of Message Brokers"""
+
+    def __init__(self, broker_conf):
+        self._servers = ','.join(x["server"]+':'+x['port'] for x in \
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker.py:50]:
```
self._servers = ','.join(x['server']+':'+x['port'] for x in 
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker.py (hunk 8 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker.py:50]: Implemented single quote in all required places, to look consistent

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        self._message_bus = message_bus
+        self._message_bus.init_client(client_type, **client_conf)
+        self._client_conf = client_conf
+
+    def _get_conf(self, key: str):
+        if key not in self._client_conf.keys():
+            raise MessageBusError(errno.EINVAL, "Invalid entry %s", key) 
+        return self._client_conf[key]
+
+    def send(self, messages: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:35]: def send(self, messages: list):

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        Parameters:
+        message_bus     An instance of message bus class.
+        producer_id     A String that represents Producer client ID.
+        message_type    This is essentially equivalent to the
+                        queue/topic name. For e.g. ["Alert"]
+        """
+        super().__init__(message_bus, client_type='producer', \
+            client_id=producer_id, message_type=message_type, method=method)
+
+    def send(self, messages: list):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:66]: Remove this interface. Not required.

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                        For e.g. ["Alert"]
+        auto_ack        Can be set to "True" or "False"
+        offset          Can be set to "earliest" (default) or "latest".
+                        ("earliest" will cause messages to be read from the beginning)
+        """
+        super().__init__(message_bus, client_type='consumer', \
+            client_id=consumer_id, consumer_group=consumer_group, \
+            message_type=message_type, auto_ack=auto_ack, offset=offset)
+
+    def receive(self) -> list:
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:93]: Remove this interface. Not required.

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        """ Receive messages from the Message Bus """
+        return super().receive()
+
+    def ack(self):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:97]: Remove this interface. Not required.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+                    raise MessageBusError(errno.ECONN, "Cant receive. %s", \
+                        msg.error())
+                else:
+                    yield ConsumerRecord(msg.topic(), msg.value(), \
+                        msg.partition(), msg.offset(), str(msg.key()))
+
+        except KeyboardInterrupt:
+            sys.stderr.write('%% Aborted by user\n')
+
+    def ack(self, consumer_id:str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:112]:
```
def ack(self, consumer_id: str):
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                    continue
+                if msg.error():
+                    raise MessageBusError(errno.ECONN, "Cant receive. %s", \
+                        msg.error())
+                else:
+                    yield ConsumerRecord(msg.topic(), msg.value(), \
+                        msg.partition(), msg.offset(), str(msg.key()))
+
+        except KeyboardInterrupt:
+            sys.stderr.write('%% Aborted by user\n')
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:110]: Remove write. You are not allowed to print onto the screen. You only can return MessageBusError

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 11 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:93]: removed

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 12 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:97]: removed

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 13 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:112]: indented

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 10 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:66]: removed

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 9 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:35]: changed type here and message_bus as well

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            producer.produce(message_type, bytes(message, 'utf-8'))
+            if method == 'sync':
+                producer.flush()
+        producer.flush()
+
+    def receive(self, consumer_id: str, timeout=0.5) -> list:
+        """ Receives list of messages to Kafka cluster(s) """
+        consumer = self._clients['consumer'][consumer_id]
+        if consumer is None:
+            raise MessageBusError(errno.EINVAL, "init_client is not called")
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:95]: raise MessageBusError(errno.EINVAL, "Consumer %s not initialized" %consumer_id)

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    def send(self, producer_id: str, message_type: str, method: str, messages: list):
+        """ Sends list of messages to Kafka cluster(s) """
+        producer = self._clients['producer'][producer_id]
+        if producer is None:
+            raise MessageBusError(errno.EINVAL, "init_client is not called")
+
+        for message in messages:
+            producer.produce(message_type, bytes(message, 'utf-8'))
+            if method == 'sync':
+                producer.flush()
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:88]: We need have only one flush remove flush if sync is not specified.

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+            consumer = Consumer(**kafka_conf)
+            consumer.subscribe(client_conf['message_type'])
+            self._clients[client_type][client_conf['client_id']] = consumer
+
+    def send(self, producer_id: str, message_type: str, method: str, messages: list):
+        """ Sends list of messages to Kafka cluster(s) """
+        producer = self._clients['producer'][producer_id]
+        if producer is None:
+            raise MessageBusError(errno.EINVAL, "init_client is not called")
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:83]: raise MessageBusError(errno.EINVAL, "Producer %s is not initialized" %producer_id)

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+        if client_type == 'producer':
+            producer = Producer(**kafka_conf)
+            self._clients[client_type][client_conf['client_id']] = producer
+
+        else:
+            for entry in ['offset', 'consumer_group', 'message_type', 'auto_ack'\
+                , 'client_id']:
+                if entry not in client_conf.keys():
+                    raise MessageBusError(errno.EINVAL, "Missing conf entry %s"\
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:68]: Comma needs to with previous line. 

raise MessageBusError(errno.EINVAL, "Missing conf entry %s", \

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        kafka_conf = {}
+        kafka_conf['bootstrap.servers'] = self._servers
+        kafka_conf['client.id'] = client_conf['client_id']
+
+        if client_type == 'producer':
+            producer = Producer(**kafka_conf)
+            self._clients[client_type][client_conf['client_id']] = producer
+
+        else:
+            for entry in ['offset', 'consumer_group', 'message_type', 'auto_ack'\
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:65]:
```
for entry in ['offset', 'consumer_group', 'message_type', 'auto_ack', \
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+import sys
+import errno
+from collections import namedtuple
+from confluent_kafka import Producer, Consumer
+from confluent_kafka.admin import AdminClient
+from cortx.utils.message_bus.error import MessageBusError
+from cortx.utils.message_bus.message_broker import MessageBroker
+
+ConsumerRecord = namedtuple("ConsumerRecord", ["message_type", "message",\
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:26]:
```
ConsumerRecord = namedtuple("ConsumerRecord", ["message_type", "message", \
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/test_message_bus/test_multiple_consumers.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+            consumer_id="sspl_sensor1", consumer_group="c3", \
+            message_type=['test_type'], auto_ack=True, offset='latest')
+
+        self.assertIsNotNone(consumer, "Consumer not found")
+        messages = consumer.receive()
+        self.assertEqual(len(list(messages)), 1000)
+        self.assertIsNotNone(messages, "Messages not found")
+        for message in messages:
+            print(message)
+        #consumer.ack()
```
Comment (ujjwalpl) [inline py-utils/test/test_message_bus/test_multiple_consumers.py:68]: Remove commented code.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 18 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:68]: taken care

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 19 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:65]: space added

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 20 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:26]: space added

File: py-utils/test/test_message_bus/test_multiple_consumers.py (hunk 21 above)
Comment (selvakumaar5496) [reply py-utils/test/test_message_bus/test_multiple_consumers.py:68]: added this code

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 17 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:83]: altered statement

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 15 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:95]: altered statement

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 16 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:88]: removed

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 14 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:110]: if i raise MessageBusError, My test case wont pass since it catches an Error

Comment (ujjwalpl) [reply py-utils/src/utils/message_bus/message_broker_collection.py:114]: write is not an option for us. This is a service and does not have tty attached. 
Your test can catch MessageBus error.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+        if client_type == 'producer':
+            producer = Producer(**kafka_conf)
+            self._clients[client_type][client_conf['client_id']] = producer
+
+        else:
+            for entry in ['offset', 'consumer_group', 'message_type', \
+                'auto_ack', 'client_id']:
+                if entry not in client_conf.keys():
+                    raise MessageBusError(errno.EINVAL, "Missing conf entry \
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:68]:
```
raise MessageBusError(errno.EINVAL, "Missing conf entry %s", entry)
```
Please note its comma. We need to pass message_format and args to the exception 
Please fix this at all the places.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+        except KeyboardInterrupt:
+            raise MessageBusError(errno.EINVAL, "Cant Recieve %s")
+
+    def ack(self, consumer_id: str):
+        """ To manually commit offset """
+        consumer = self._clients['consumer'][consumer_id]
+        if consumer is None:
+            raise MessageBusError(errno.EINVAL, "Consumer %s is not \
+                initialized", consumer_id)
+        consumer.commit()
```
Comment (ajaysrivas) [inline py-utils/src/utils/message_bus/message_broker_collection.py:114]: async=false

Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py]: Hi Ajay. Implemented async on commit

**Review by ajaysrivas — APPROVED** [review]

**Pull request closed without merging**